// antd-mobile/lib/locale-provider/en_US.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _en_US = require('../date-picker-view/locale/en_US');

var _en_US2 = _interopRequireDefault(_en_US);

var _en_US3 = require('../date-picker/locale/en_US');

var _en_US4 = _interopRequireDefault(_en_US3);

var _en_US5 = require('../input-item/locale/en_US');

var _en_US6 = _interopRequireDefault(_en_US5);

var _en_US7 = require('../pagination/locale/en_US');

var _en_US8 = _interopRequireDefault(_en_US7);

var _en_US9 = require('../picker/locale/en_US');

var _en_US10 = _interopRequireDefault(_en_US9);

var _en_US11 = require('../search-bar/locale/en_US');

var _en_US12 = _interopRequireDefault(_en_US11);

var _en_US13 = require('../menu/locale/en_US');

var _en_US14 = _interopRequireDefault(_en_US13);

var _en_US15 = require('../pull-to-refresh/locale/en_US');

var _en_US16 = _interopRequireDefault(_en_US15);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

exports['default'] = {
    locale: 'en',
    Pagination: _en_US8['default'],
    DatePicker: _en_US4['default'],
    DatePickerView: _en_US2['default'],
    InputItem: _en_US6['default'],
    Picker: _en_US10['default'],
    SearchBar: _en_US12['default'],
    Menu: _en_US14['default'],
    PullToRefresh: _en_US16['default']
};
module.exports = exports['default'];
