// antd-mobile/lib/locale-provider/ru_RU.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _ru_RU = require('../date-picker-view/locale/ru_RU');

var _ru_RU2 = _interopRequireDefault(_ru_RU);

var _ru_RU3 = require('../date-picker/locale/ru_RU');

var _ru_RU4 = _interopRequireDefault(_ru_RU3);

var _ru_RU5 = require('../input-item/locale/ru_RU');

var _ru_RU6 = _interopRequireDefault(_ru_RU5);

var _ru_RU7 = require('../pagination/locale/ru_RU');

var _ru_RU8 = _interopRequireDefault(_ru_RU7);

var _ru_RU9 = require('../picker/locale/ru_RU');

var _ru_RU10 = _interopRequireDefault(_ru_RU9);

var _ru_RU11 = require('../search-bar/locale/ru_RU');

var _ru_RU12 = _interopRequireDefault(_ru_RU11);

var _ru_RU13 = require('../menu/locale/ru_RU');

var _ru_RU14 = _interopRequireDefault(_ru_RU13);

var _ru_RU15 = require('../pull-to-refresh/locale/ru_RU');

var _ru_RU16 = _interopRequireDefault(_ru_RU15);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

exports['default'] = {
    locale: 'ru',
    Pagination: _ru_RU8['default'],
    DatePicker: _ru_RU4['default'],
    DatePickerView: _ru_RU2['default'],
    InputItem: _ru_RU6['default'],
    Picker: _ru_RU10['default'],
    SearchBar: _ru_RU12['default'],
    Menu: _ru_RU14['default'],
    PullToRefresh: _ru_RU16['default']
};
module.exports = exports['default'];
