// antd-mobile/lib/locale-provider/sv_SE.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _sv_SE = require('../date-picker-view/locale/sv_SE');

var _sv_SE2 = _interopRequireDefault(_sv_SE);

var _sv_SE3 = require('../date-picker/locale/sv_SE');

var _sv_SE4 = _interopRequireDefault(_sv_SE3);

var _sv_SE5 = require('../input-item/locale/sv_SE');

var _sv_SE6 = _interopRequireDefault(_sv_SE5);

var _sv_SE7 = require('../pagination/locale/sv_SE');

var _sv_SE8 = _interopRequireDefault(_sv_SE7);

var _sv_SE9 = require('../picker/locale/sv_SE');

var _sv_SE10 = _interopRequireDefault(_sv_SE9);

var _sv_SE11 = require('../search-bar/locale/sv_SE');

var _sv_SE12 = _interopRequireDefault(_sv_SE11);

var _sv_SE13 = require('../menu/locale/sv_SE');

var _sv_SE14 = _interopRequireDefault(_sv_SE13);

var _sv_SE15 = require('../pull-to-refresh/locale/sv_SE');

var _sv_SE16 = _interopRequireDefault(_sv_SE15);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

exports['default'] = {
    locale: 'sv',
    Pagination: _sv_SE8['default'],
    DatePicker: _sv_SE4['default'],
    DatePickerView: _sv_SE2['default'],
    InputItem: _sv_SE6['default'],
    Picker: _sv_SE10['default'],
    SearchBar: _sv_SE12['default'],
    Menu: _sv_SE14['default'],
    PullToRefresh: _sv_SE16['default']
};
module.exports = exports['default'];
